frontend/public/components/routes.tsx

```diff
@@ -307,14 +312,297 @@ const filters = [{
+  const portOptions = ports.reduce((acc, { name, port, targetPort, protocol }) => {
+    acc[name] = <React.Fragment>{port} &rarr; {targetPort} ({protocol})</React.Fragment>;
+    return acc;
+  }, {});
+
+  return portOptions;
+};
+
+export class CreateRoute extends React.Component<null, CreateRouteState> {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      formData: {
+        name: '',
+        hostname: '',
+        path: '',
+        service: '',
+        targetPort: '',
+        error: '',
+      },
+      toggleData: {
+        alternate: false,
+        secure: false,
+      },
+      loaded: false,
+      inProgress: false,
+      namespace: getActiveNamespace(),
+      services: [],
+      labels: {},
+      portOptions: {},
+    };
+  }
+
+  componentDidMount() {
+    k8sGet(ServiceModel, null, this.state.namespace)
+      .then(services => this.setState({
+        services: services.items,
+        loaded: true,
+      }))
+      .catch((err) => {
+        if (err.response.status !== 403) {
+          const errorMessage = err.message || 'Could not load services.';
+          this.setState({
+            formData: {
+              ...this.state.formData,
+              error: errorMessage
+            }
+          });
+        }
+      });
+  }
+
+  handleChange: React.ReactEventHandler<HTMLInputElement> = event => {
+    this.setState({
+      formData: {
+        ...this.state.formData,
+        [event.currentTarget.name]: event.currentTarget.value
+      }
+    });
+  }
+
+  changeService = (serviceName: string) => {
+    const selectedService = _.find(this.state.services, { metadata: { name: serviceName } });
+    const portOptions = getServicePorts(selectedService);
+    const labels = selectedService.metadata.labels;
+    this.setState({
+      formData: {
+        ...this.state.formData,
+        service: serviceName,
+      },
+      labels,
+      portOptions,
+    });
+  }
+
+  changeTargetPort = (portName: string) => {
+    this.setState({
+      formData: {
+        ...this.state.formData,
+        targetPort: portName
+      }
+    });
+  }
+
+  toggleSection: React.ReactEventHandler<HTMLInputElement> = event => {
+    this.setState({
+      toggleData: {
+        ...this.state.toggleData,
+        [event.currentTarget.name]: event.currentTarget.checked
+      }
+    });
+  }
+
+  save = event => {
+    const {
+      name,
+      hostname,
+      path,
+      service,
+      targetPort,
+    } = this.state.formData;
+    const {
+      namespace,
+      labels,
+    } = this.state;
+    const route = {
+      kind: 'Route',
+      apiVersion: 'route.openshift.io/v1',
+      metadata: {
+        name,
+        namespace,
+        labels,
+      },
+      spec: {
+        to: {
+          kind: 'Service',
+          name: service
+        },
+        host: hostname,
+        path,
+        port: {
+          targetPort,
+        }
+      }
+    };
+
+    event.preventDefault();
```

Comment:
nit: I usually make this the first line of the handler. (It would be easy for someone to add an early return later and accidentally skip prevent default.)